keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -0,0 +1,1253 @@
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### New PodConditions
+
+The already existing `status.conditions` field in Pod will be used by kubernetes
+components (such as: kube-scheduler, kubelet pod-gc) to append a dedicated
+condition when they send the delete operation.
+
+During the implementation process we are going to review the places where the
+pod delete requests are issued to modify the code to also append a meaningful
+condition with dedicated `Type`, `Reason` and `Message` fields based on the
+invocation context.
+
+### JobSpec API
+
+We extend the Job API in order to allow to apply different actions depending
+on the conditions associated with the pod failure.
+
+```golang
+// BackoffPolicyAction specifies how a Pod failure is handled.
+type BackoffPolicyAction string
+
+const (
+
+	// This is an action which might be taken on pod failure - mark the
+	// pod's job as Failed and terminate all running pods.
+	BackoffActionTerminate BackoffPolicyAction = "Terminate"
+
+	// This is an action which might be taken on pod failure - the pod will be
+	// restarted and the counter for .backoffLimit will not be incremented.
+	BackoffActionIgnore BackoffPolicyAction = "Ignore"
+)
+
+type BackoffPolicyOnExitCodeInterval struct {
+	MinInclusive int
+	MaxInclusive int
+}
+
+type BackoffPolicyOnExitCodeRequirement struct {
+	// Restricts the check for exit codes to only apply to the container with
+	// the specified name. When empty the rule applies to all containers
+	// +optional
+	ContainerName *string
+
+	// Specifies the set of intervals. The requirement is satisfied when the
+	// actual exit code is in any of the intervals
+	// +optional
+	InRange *[]BackoffPolicyOnExitCodeInterval
+
+	// Specifies the set of intervals. The requirement is satisfied when the
+	// actual exit code is not in any of the intervals
+	// +optional
+	NotInRange *[]BackoffPolicyOnExitCodeInterval
+
+	// Specifies the set of exit codes. The requirement is satisfied when the
+	// actual exit code is in the set
+	// +optional
+	In *[]int
+
+	// Specifies the set of exit codes. The requirement is satisfied when the
+	// actual exit code is not in the set
+	// +optional
+	NotIn *[]int
+}
+
+type BackoffPolicyOnPodConditionOperator string
+
+const (
+	BackoffPolicyOnPodConditionOperatorIn    BackoffPolicyOnPodConditionOperator = "In"
+	BackoffPolicyOnPodConditionOperatorNotIn BackoffPolicyOnPodConditionOperator = "NotIn"
+)
+
+type BackoffPolicyOnPodConditionsRequirement struct {
+
+	// Represents the name of the field in the pod condition
+	Key string
```

Comment:
We should always be matching with a Type.

Reply:
Just by looking at how the API is getting complicated, I really don't think we should be looking at reasons. See the documentation:

```
	// Unique, one-word, CamelCase reason for the condition's last transition.
	// +optional
	Reason string
```

And all the possible reasons are not really documented. So I vote we ignore them in this API.

Maybe we can do something like this:

```golang
type BackoffPolicyOnPodConditionsRequirement struct {
  // Could be AnyOf, AllOf
  Operator BackoffPolicyOnPodConditionOperator
  Conditions []PodConditionMatcher
}

type PodConditionMatcher {
  Type PodConditionType
  // optional, true by default
  Status *PodConditionStatus
```

And if we later decide that comparing reasons is necessary, we can expand the API by adding `Reasons []string`

Reply:
> We should always be matching with a Type.

I agree with this. Without a specific condition type, the meaning of the status field is undefined.

Reply:
If the intent is to retry only on infrastructure errors, that is not a well-defined set of conditions.

Currently, that includes things like Evicted, Orphaned, Preempted, NodeShutdown, but that is not necessarily a bounded set, and could grow in the future (just as NodeShutdown was added a couple releases ago).

We don't want to make job authors responsible for discovering/enumerating all the various condition types related to infrastructure pod failures. If they do, that means they are likely to miss rarely encountered conditions, and if we add new infrastructure conditions in the future, previously "correct" job specs would become incorrect.

A couple things could help here:
1. make the retry policy only go in the `Terminate` direction... it should opt out of retrying for specific conditions or exit codes (e.g. "skip retrying on x,y,z")... that fails in a safer direction if a rare condition is accidentally missed
2. it would be more user-friendly if a single condition identified when a pod terminated "normally" rather than because of external reasons (infrastructure of pod deletion)

Reply:
Things we seem to agree:
- We need an explicit field to match the PodCondition.Type
- Since conditions already have an enumeration problem, we probably don't want to make it worse by paying attention to the "reason", which is more free form than type.

> A couple things could help here:

I think either 1 or 2 are sufficient to prevent authors from having to discover infrastructure errors.
If we go with 1, a provider can only express "retry up to `backoffLimit` infrastructure errors, but terminate on first user error". But it can't express "retry indefinitely on infrastructure errors, but retry up to `backoffLimit` user errors"
Doing just 2 allows to express both.